src/Internal/RestApiUtil.php

```diff
@@ -0,0 +1,195 @@
+
+namespace Automattic\WooCommerce\Internal;
+
+/**
+ * Helper methos for the REST API.
+ *
+ * Class ApiUtil
+ *
+ * @package Automattic\WooCommerce\Internal
+ */
+class RestApiUtil {
+
+	/**
+	 * Converts a create refund request from the public API format:
+	 *
+	 * [
+	 *   "reason" => "",
+	 *   "api_refund" => "x",
+	 *   "line_items" => [
+	 *     "id" => "111",
+	 *     "quantity" => 222,
+	 *     "refund_total" => 333,
+	 *     "refund_tax" => [
+	 *       [
+	 *          "id": "444",
+	 *          "refund_total": 555
+	 *       ],...
+	 *   ],...
+	 * ]
+	 *
+	 * ...to the internally used format:
+	 *
+	 * [
+	 *   "reason" => null,     (if it's missing or any empty value, set as null)
+	 *   "api_refund" => true, (if it's missing or non-bool, set as "true")
+	 *   "line_items" => [     (convert sequential array to associative based on "id")
+	 *     "111" => [
+	 *       "qty" => 222,     (rename "quantity" to "qty")
+	 *       "refund_total" => 333,
+	 *       "refund_tax" => [ (convert sequential array to associative based on "id" and "refund_total)
+	 *         "444" => 555,...
+	 *       ],...
+	 *   ]
+	 * ]
+	 *
+	 * It also calculates the amount if missing and whenever possible, see maybe_calculate_refund_amount_from_line_items.
+	 *
+	 * The conversion is done in a way that if the request is already in the internal format,
+	 * then nothing is changed for compatibility. For example, if the line items array
+	 * is already an associative array or any of its elements
+	 * is missing the "id" key, then the entire array is left unchanged.
+	 * Same for the "refund_tax" array inside each line item.
+	 *
+	 * @param \WP_REST_Request $request The request to adjust.
+	 */
+	public static function adjust_create_refund_request_parameters( \WP_REST_Request &$request ) {
+		if ( empty( $request['reason'] ) ) {
+			$request['reason'] = null;
+		}
+
+		if ( ! is_bool( $request['api_refund'] ) ) {
+			$request['api_refund'] = true;
+		}
+
+		if ( empty( $request['line_items'] ) ) {
+			$request['line_items'] = array();
+		} else {
+			$request['line_items'] = self::adjust_line_items_for_create_refund_request( $request['line_items'] );
+		}
+
+		self::maybe_calculate_refund_amount_from_line_items( $request );
+	}
+
+	/**
+	 * Calculate and set the "amount" parameter in the request based on the amounts found in line items.
+	 * This will ONLY happen if ALL of the following is true:
+	 *
+	 * - The request doesn't have an "amount" parameter already.
+	 * - "line_items" in the request is a non-empty array.
+	 * - All line items have a "refund_total" field with a numeric value.
+	 * - All values inside "refund_tax" in all line items are a numeric value.
+	 *
+	 * The request is assumed to be in internal format already.
+	 *
+	 * @param \WP_REST_Request $request The request to maybe calculate the total amount for.
+	 */
+	private static function maybe_calculate_refund_amount_from_line_items( &$request ) {
+		if ( isset( $request['amount'] ) ) {
```

Comment:
shouldn't we give preference to line item amount if is present? what happens when both are provided but don't match?

Reply:
The idea is that if the total `amount` is present in the request we assume that the client knows what he's doing and so we don't calculate anything, the provided amount is used even if it wouldn't match the calculation based on `refund_total`s. That's how it works now and I understand that we should keep this behavior for compatibility.

Reply:
I tested this out, with amounts not matching:

1. When done from the API, in master, we just ignore line items altogether.

2. In UI, we don't allow to specify refund amount at all. It has to be entered via line items, and there fore takes precedence.

I guess both ways are correct then. Going approve this PR then.